File: bindings/i18n/i18n.go
package i18n

import (
	"fmt"

	core "dappco.re/go/core"
	"dappco.re/go/py/bindings/typemap"
	"dappco.re/go/py/runtime"
)

type translator interface {
	Translate(messageID string, args ...any) core.Result
	SetLanguage(lang string) error
	Language() string
	AvailableLanguages() []string
}

type handle struct {
	locales    []any
	locale     string
	translator translator
}

// Register exposes Core i18n helpers.
//
//	i18n.Register(interpreter)
func Register(interpreter *runtime.Interpreter) error {
	return interpreter.RegisterModule(runtime.Module{
		Name:          "core.i18n",
		Documentation: "Locale and translation helpers for CorePy",
		Functions: map[string]runtime.Function{
			"new":                 newI18n,
			"add_locales":         addLocales,
			"locales":             locales,
			"set_translator":      setTranslator,
			"translator":          translatorValue,
			"translate":           translate,
			"set_language":        setLanguage,
			"language":            language,
			"available_languages": availableLanguages,
		},
	})
}

func newI18n(arguments ...any) (any, error) {
	return &handle{}, nil
}

func addLocales(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.add_locales")
	if err != nil {
		return nil, err
	}
	item.locales = append(item.locales, arguments[1:]...)
	return item, nil
}

func locales(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.locales")
	if err != nil {
		return nil, err
	}
	result := make([]any, len(item.locales))
	copy(result, item.locales)
	return result, nil
}

func setTranslator(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.set_translator")
	if err != nil {
		return nil, err
	}
	if len(arguments) < 2 || arguments[1] == nil {
		item.translator = nil
		return item, nil
	}
	translatorValue, ok := arguments[1].(translator)
	if !ok {
		return nil, fmt.Errorf("core.i18n.set_translator expected translator, got %T", arguments[1])
	}
	item.translator = translatorValue
	if item.locale != "" {
		if err := item.translator.SetLanguage(item.locale); err != nil {
			return nil, err
		}
	}
	return item, nil
}

func translatorValue(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.translator")
	if err != nil {
		return nil, err
	}
	if item.translator == nil {
		return nil, nil
	}
	return item.translator, nil
}

func translate(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.translate")
	if err != nil {
		return nil, err
	}
	messageID, err := typemap.ExpectString(arguments, 1, "core.i18n.translate")
	if err != nil {
		return nil, err
	}
	if item.translator == nil {
		return messageID, nil
	}
	return typemap.ResultValue(item.translator.Translate(messageID, arguments[2:]...), "core.i18n.translate")
}

func setLanguage(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.set_language")
	if err != nil {
		return nil, err
	}
	lang, err := typemap.ExpectString(arguments, 1, "core.i18n.set_language")
	if err != nil {
		return nil, err
	}
	if lang == "" {
		return item, nil
	}
	item.locale = lang
	if item.translator != nil {
		if err := item.translator.SetLanguage(lang); err != nil {
			return nil, err
		}
	}
	return item, nil
}

func language(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.language")
	if err != nil {
		return nil, err
	}
	if item.locale != "" {
		return item.locale, nil
	}
	if item.translator != nil && item.translator.Language() != "" {
		return item.translator.Language(), nil
	}
	return "en", nil
}

func availableLanguages(arguments ...any) (any, error) {
	item, err := expectHandle(arguments, 0, "core.i18n.available_languages")
	if err != nil {
		return nil, err
	}
	if item.translator == nil {
		return []string{"en"}, nil
	}
	return item.translator.AvailableLanguages(), nil
}

func expectHandle(arguments []any, index int, functionName string) (*handle, error) {
	if index >= len(arguments) {
		return nil, fmt.Errorf("%s expected argument %d", functionName, index)
	}
	value, ok := arguments[index].(*handle)
	if !ok {
		return nil, fmt.Errorf("%s expected i18n handle, got %T", functionName, arguments[index])
	}
	return value, nil
}

File: go/py/core/i18n.py
"""Locale and translation helpers.

from core import i18n

messages = i18n.new()
"""

from __future__ import annotations

import builtins
from typing import Any, Protocol


class Translator(Protocol):
    def translate(self, message_id: str, *args: Any) -> Any: ...
    def set_language(self, lang: str) -> None: ...
    def language(self) -> str: ...
    def available_languages(self) -> list[str]: ...


class I18n:
    """Locale collection plus optional translator dispatch.

    messages = i18n.I18n()
    """

    def __init__(self) -> None:
        self._locales: list[Any] = []
        self._locale = ""
        self._translator: Translator | None = None

    def add_locales(self, *mounts: Any) -> None:
        """Append locale mounts.

        messages.add_locales("locales")
        """

        self._locales.extend(mounts)

    def locales(self) -> list[Any]:
        """Return collected locale mounts.

        messages.locales()
        """

        return builtins.list(self._locales)

    def set_translator(self, translator: Translator | None) -> None:
        """Register a translator implementation.

        messages.set_translator(translator)
        """

        self._translator = translator
        if translator is not None and self._locale:
            translator.set_language(self._locale)

    def translator(self) -> Translator | None:
        """Return the registered translator.

        messages.translator()
        """

        return self._translator

    def translate(self, message_id: str, *args: Any) -> Any:
        """Translate a message or return the key as-is.

        messages.translate("hello")
        """

        if self._translator is None:
            return message_id
        return self._translator.translate(message_id, *args)

    def set_language(self, lang: str) -> None:
        """Set the active language.

        messages.set_language("de")
        """

        if lang == "":
            return
        self._locale = lang
        if self._translator is not None:
            self._translator.set_language(lang)

    def language(self) -> str:
        """Return the active language or `en`.

        messages.language()
        """

        if self._locale:
            return self._locale
        if self._translator is not None:
            value = self._translator.language()
            if value:
                return value
        return "en"

    def available_languages(self) -> list[str]:
        """Return available language codes.

        messages.available_languages()
        """

        if self._translator is None:
            return ["en"]
        return builtins.list(self._translator.available_languages())


def new() -> I18n:
    """Create an I18n handle.

    i18n.new()
    """

    return I18n()


def add_locales(i18n_value: I18n, *mounts: Any) -> I18n:
    """Append locale mounts and return the handle.

    i18n.add_locales(messages, "locales")
    """

    i18n_value.add_locales(*mounts)
    return i18n_value


def locales(i18n_value: I18n) -> list[Any]:
    """Return collected locale mounts.

    i18n.locales(messages)
    """

    return i18n_value.locales()


def set_translator(i18n_value: I18n, translator: Translator | None) -> I18n:
    """Register a translator and return the handle.

    i18n.set_translator(messages, translator)
    """

    i18n_value.set_translator(translator)
    return i18n_value


def translator(i18n_value: I18n) -> Translator | None:
    """Return the registered translator.

    i18n.translator(messages)
    """

    return i18n_value.translator()


def translate(i18n_value: I18n, message_id: str, *args: Any) -> Any:
    """Translate a message or return the key.

    i18n.translate(messages, "hello")
    """

    return i18n_value.translate(message_id, *args)


def set_language(i18n_value: I18n, lang: str) -> I18n:
    """Set the active language and return the handle.

    i18n.set_language(messages, "de")
    """

    i18n_value.set_language(lang)
    return i18n_value


def language(i18n_value: I18n) -> str:
    """Return the active language.

    i18n.language(messages)
    """

    return i18n_value.language()


def available_languages(i18n_value: I18n) -> list[str]:
    """Return available language codes.

    i18n.available_languages(messages)
    """

    return i18n_value.available_languages()


__all__ = [
    "I18n",
    "Translator",
    "add_locales",
    "available_languages",
    "language",
    "locales",
    "new",
    "set_language",
    "set_translator",
    "translate",
    "translator",
]
